klippy/extras/bltouch.py

```diff
@@ -53,54 +59,96 @@ def __init__(self, config):
+        else:
+            self.next_cmd_time = print_time
+    def send_cmd(self, cmd, duration=MIN_CMD_TIME):
+        self.mcu_pwm.set_pwm(self.next_cmd_time, Commands[cmd] / SIGNAL_PERIOD)
+        # Translate duration to ticks to avoid any secondary mcu clock skew
+        mcu = self.mcu_pwm.get_mcu()
+        cmd_clock = mcu.print_time_to_clock(self.next_cmd_time)
+        cmd_clock += mcu.seconds_to_clock(max(duration, MIN_CMD_TIME))
+        self.next_cmd_time = mcu.clock_to_print_time(cmd_clock)
+        return self.next_cmd_time
+    def verify_state(self, check_start_time, check_end_time, triggered, msg):
+        # Perform endstop check to verify bltouch reports desired state
         prev_positions = [s.get_commanded_position()
                           for s in self.mcu_endstop.get_steppers()]
-        self.mcu_endstop.home_start(home_time, ENDSTOP_SAMPLE_TIME,
-                                    ENDSTOP_SAMPLE_COUNT, ENDSTOP_REST_TIME)
+        self.mcu_endstop.home_start(check_start_time, ENDSTOP_SAMPLE_TIME,
+                                    ENDSTOP_SAMPLE_COUNT, ENDSTOP_REST_TIME,
+                                    triggered=triggered)
         try:
-            self.mcu_endstop.home_wait(home_time + MIN_CMD_TIME)
+            self.mcu_endstop.home_wait(check_end_time)
         except self.mcu_endstop.TimeoutError as e:
-            raise homing.EndstopError("BLTouch sensor test failed")
+            raise homing.EndstopError("BLTouch failed to %s" % (msg,))
         for s, pos in zip(self.mcu_endstop.get_steppers(), prev_positions):
             s.set_commanded_position(pos)
+    def raise_probe(self):
+        self.send_cmd('reset')
+        check_start_time = self.send_cmd('pin_up', duration=self.pin_move_time)
+        check_end_time = self.send_cmd(None)
+        self.verify_state(check_start_time, check_end_time,
+                          self.pin_up_triggered, "raise probe")
+    def test_sensor(self):
+        if not self.pin_up_touch_triggered or self.pin_up_triggered:
+            # Nothing to test
+            return
+        toolhead = self.printer.lookup_object('toolhead')
+        print_time = toolhead.get_last_move_time()
+        if print_time < self.next_test_time:
+            self.next_test_time = print_time + TEST_TIME
+            return
+        # Raise the bltouch probe and test if probe is raised
+        self.sync_print_time()
+        check_start_time = self.send_cmd('reset', duration=self.pin_move_time)
+        check_end_time = self.send_cmd('touch_mode')
+        self.send_cmd(None)
+        self.verify_state(check_start_time, check_end_time, True,
+                          "verify sensor state")
         # Test was successful
-        self.next_test_time = home_time + TEST_TIME
-        toolhead.reset_print_time(home_time + 2. * MIN_CMD_TIME)
+        self.next_test_time = check_end_time + TEST_TIME
+        self.sync_print_time()
```

Comment:
Ok, perhaps it detects reversed endstop polarity or a missing pull up resistor, when the cable is disconnected. So it might not be totally useless.

Reply:
The test is intended as a basic wiring / configuration / bltouch-not-totally-fried test.  The goal is to command the bltouch and verify that the signal line reports triggered, then command the bltouch and verify the sensor line reports not-triggered.  It is believed that this triggered-then-not-triggered sequence will verify that both the control and sensor pins are sane.  It is desirable to do that prior to moving the head towards the bed.

It is understood that a geniune bltouch has self test capabilities, but that's only useful if the bltouch is operating nominally and that klipper is effectively communicating with the bltouch.

Situations 2, 3, 4 above do not pass this test - all of these get flagged by the second, "not triggered", part of the test.  The "not triggered" test is implicitly done by verifying that the stepper motors move some distance during the actual homing sequence.  (If the sensor pin stayed in a triggered state then the homing sequence will end without any movement on the steppers.)

Reply:
Well they pass *this* test and the homing operation is allowed to start.

But yes, you are right it's probably useful to detect the two cases I mentioned above and a third one that came to my mind, even if a well built probe should be able to detect all those itself:
1. Reversed endstop pin polarity
2. Disconnected endstop cable in addition to not having a pull up resistor configured
3. Disconnected signal cable, but powered probe

So, yes the test can remain, especially since it's not run very often, so I won't bug you about this anymore, my issue was more about this being the *only* test there was.